File: src/main/java/com/todoist/pojo/Person.java
package com.todoist.pojo;

import java.util.regex.Pattern;

public class Person extends TodoistObject {
    /* Matches leading and trailing spaces and special characters commonly found in name fields. */
    private static final Pattern ESCAPE_PATTERN = Pattern.compile("^\\s+|\\s+$|[\\(\\)\\.,\\-_\\[\\]'\"]");

    private static final int[] LIGHT_AVATAR_COLORS = {
            0XFFE9952C,
            0XFFE16B2D,
            0XFFD84B40,
            0XFFE8435A,
            0XFFE5198A,
            0XFFAD3889,
            0XFF86389C,
            0XFF98BE2F,
            0XFF5D9D50,
            0XFF5F9F85,
            0XFF5BBCB6,
            0XFF32A3BF,
            0XFF2BAFEB,
            0XFF2D88C3,
            0XFF3863CC,
            0XFF5E5E5E
    };

    private static final int[] DARK_AVATAR_COLORS = {
            0XFFFCC652,
            0XFFE9952C,
            0XFFE16B2D,
            0XFFD84B40,
            0XFFE8435A,
            0XFFE5198A,
            0XFFAD3889,
            0XFFA8A8A8,
            0XFF98BE2F,
            0XFF5D9D50,
            0XFF5F9F85,
            0XFF5BBCB6,
            0XFF32A3BF,
            0XFF2BAFEB,
            0XFF2D88C3
    };

    private String email;
    private String fullName;
    private String imageId;

    public Person(long id, String email, String fullName, String imageId, boolean deleted) {
        super(id, deleted);
        this.email = email;
        this.fullName = fullName;
        this.imageId = imageId;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getFullName() {
        return fullName;
    }

    public void setFullName(String fullName) {
        this.fullName = fullName;
    }

    public String getImageId() {
        return imageId;
    }

    public void setImageId(String imageId) {
        this.imageId = imageId;
    }

    public int getDefaultAvatarColorInt(boolean useLightColors) {
        return getDefaultAvatarColorInt(email, useLightColors);
    }

    public String getDefaultAvatarText() {
        return getDefaultAvatarText(fullName);
    }

    public static int getDefaultAvatarColorInt(String email, boolean useLightColors) {
        if (email != null) {
            int atIndex = email.indexOf("@");
            if (atIndex > 0) {
                int[] colors = getAvatarColors(useLightColors);
                return colors[(email.charAt(0) + email.charAt(atIndex - 1)) % colors.length];
            }
        }

        return 0xFF000000;
    }

    private static int[] getAvatarColors(boolean useLightColors) {
        return useLightColors ? LIGHT_AVATAR_COLORS : DARK_AVATAR_COLORS;
    }

    public static String getDefaultAvatarText(String fullName) {
        if (fullName != null) {
            String[] names = ESCAPE_PATTERN.matcher(fullName).replaceAll("").split("\\s+");

            if (names.length > 0) {
                StringBuilder initials = new StringBuilder(2);

                String firstName = names[0];
                if (firstName.length() > 0) {
                    initials.append(Character.toUpperCase(firstName.charAt(0)));
                }

                if (names.length > 1) {
                    String lastName = names[names.length - 1];
                    if (lastName.length() > 0) {
                        initials.append(Character.toUpperCase(lastName.charAt(0)));
                    }
                }

                if (initials.length() > 0) {
                    return initials.toString();
                }
            }
        }

        return "?";
    }
}

File: src/main/java/com/todoist/pojo/Person.kt
package com.todoist.pojo

open class Person(
    id: Long,
    open var email: String,
    open var fullName: String = "",
    open var imageId: String?,
    isDeleted: Boolean
) : Model(id, isDeleted) {
    val defaultAvatarText get() = getDefaultAvatarText(fullName, email)

    fun getDefaultAvatarColorInt(useLightColors: Boolean) =
        getDefaultAvatarColorInt(email, useLightColors)

    companion object {
        // Matches special characters commonly found in name fields.
        private val ESCAPE_PATTERN = Regex("""[().,\-_\[\]'"]""")

        private val LIGHT_AVATAR_COLORS = uintArrayOf(
            0XFFE9952Cu,
            0XFFE16B2Du,
            0XFFD84B40u,
            0XFFE8435Au,
            0XFFE5198Au,
            0XFFAD3889u,
            0XFF86389Cu,
            0XFF98BE2Fu,
            0XFF5D9D50u,
            0XFF5F9F85u,
            0XFF5BBCB6u,
            0XFF32A3BFu,
            0XFF2BAFEBu,
            0XFF2D88C3u,
            0XFF3863CCu,
            0XFF5E5E5Eu
        ).toIntArray()

        private val DARK_AVATAR_COLORS = uintArrayOf(
            0XFFFCC652u,
            0XFFE9952Cu,
            0XFFE16B2Du,
            0XFFD84B40u,
            0XFFE8435Au,
            0XFFE5198Au,
            0XFFAD3889u,
            0XFFA8A8A8u,
            0XFF98BE2Fu,
            0XFF5D9D50u,
            0XFF5F9F85u,
            0XFF5BBCB6u,
            0XFF32A3BFu,
            0XFF2BAFEBu,
            0XFF2D88C3u
        ).toIntArray()

        @JvmStatic
        fun getDefaultAvatarColorInt(email: String, useLightColors: Boolean): Int {
            val atIndex = email.indexOf('@')
            if (atIndex <= 0) return 0xFF000000u.toInt()

            val colors = if (useLightColors) LIGHT_AVATAR_COLORS else DARK_AVATAR_COLORS
            val index = (email[0].code + email[atIndex - 1].code) % colors.size
            return colors[index]
        }

        @JvmStatic
        fun getDefaultAvatarText(fullName: String?, email: String?): String {
            val escapedName = fullName?.replace(ESCAPE_PATTERN, "")?.trim().orEmpty()
            if (escapedName.isBlank()) {
                return getEmailInitials(email)
            }

            val first = escapedName[0].uppercaseChar()
            val lastSpace = escapedName.indexOfLast { it.isWhitespace() }
            val last =
                if (lastSpace != -1) escapedName[lastSpace + 1].uppercaseChar().toString() else ""
            return first + last
        }

        private fun getEmailInitials(email: String?): String {
            val validEmail = email ?: "?"
            return buildString {
                val emailInitials = validEmail.substringBefore("@")
                    .split(".")
                    .filter { it.isNotEmpty() }
                    .map { emailPart -> emailPart.first().uppercase() }
                append(emailInitials.firstOrNull().orEmpty())
                if (emailInitials.size > 1) {
                    append(emailInitials.lastOrNull().orEmpty())
                }
            }
        }
    }
}
